kv/gc: resolve abandoned replicated locks during MVCC GC

nvb opened the pull request:
Fixes #111215.

This PR teaches MVCC GC to resolve abandoned replicated locks of any strength, instead of just those with Intent strength. The change itself is a one-liner to adjust the MatchMinStr filter supplied to the LockTableIterator. It then includes a bit of type refactoring to use `roachpb.Lock` in place of `roachpb.Intent`.

In the second commit, this PR then performs a series of renames from "intent" to "lock" where appropriate in the `pkg/kv/kvserver/gc` package. However, it does not change all references to "intent". Notably, references to the "intent resolver" are retained, because that structure has not yet been renamed to account for its more general role of resolving locks with any strength.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/111683)
<!-- Reviewable:end -->

nvb committed:
kv/gc: resolve abandoned replicated locks during MVCC GC

Fixes #111215.

This commit teaches MVCC GC to resolve abandoned replicated locks of any
strength, instead of just those with Intent strength. The change itself
is a one-liner to adjust the MatchMinStr filter supplied to the
LockTableIterator. It then includes a bit of type refactoring to use
roachpb.Lock in place of roachpb.Intent.

The commit doesn't go all the way with the renames from "intent" to
"lock". The more broad rename will be performed in the next commit,
which does not change any behavior.

Release note: None

nvb committed:
kv/gc: rename "intent" to "lock" in MVCC GC

This commit performs a series of renames from "intent" to "lock" where
appropriate in the `pkg/kv/kvserver/gc` package.

It also adds the following aliases for cluster settings:
```
kv.gc.lock_age_threshold           -> kv.gc.intent_age_threshold
kv.gc.lock_cleanup_batch_size      -> kv.gc.intent_cleanup_batch_size
kv.gc.lock_cleanup_batch_byte_size -> kv.gc.intent_cleanup_batch_byte_size
```

It does not change any behavior.

It also does not change all references to "intent". Notably, references
to the "intent resolver" are retained, because that structure has not
yet been renamed to account for its more general role of resolving locks
with any strength.

Epic: None
Release note: None

aliher1911 commented on pkg/cli/debug.go:
```diff
@@ -793,13 +793,13 @@ func runDebugGCCmd(cmd *cobra.Command, args []string) error {
 		var err error
-		if intentAgeThreshold, err = parsePositiveDuration(args[3]); err != nil {
+		if lockAgeThreshold, err = parsePositiveDuration(args[3]); err != nil {
 			return errors.Wrapf(err, "unable to parse %v as intent age threshold", args[3])
```
`"unable to parse %v as lock age threshold"`

nvanbenschoten replied:
Done.

aliher1911 reviewed (approved):
Looks good.

We might want to create a task to add locks to gc_randomized_tests. There's a `dataDistribution` type which is responsible for populating test data in engine which currently puts intents if txn is generated. Would be nice to also have locks there.

nvanbenschoten commented on pkg/kv/kvserver/gc/gc.go:
```diff
@@ -103,31 +103,33 @@ var TxnCleanupThreshold = settings.RegisterDurationSetting(
+	"if non zero, gc will split found locks into batches of this size when trying to resolve them",
 	5000,
 	settings.NonNegativeInt,
+	settings.WithName("kv.gc.lock_cleanup_batch_size"),
```
Done.

nvanbenschoten commented:
TFTR!

bors r=aliher1911

> We might want to create a task to add locks to gc_randomized_tests.

Agreed. I opened https://github.com/cockroachdb/cockroach/issues/111857 to track this.
